File: src/main/java/com/nadmm/airports/TabPagerActivityBase.java
/*
 * FlightIntel for Pilots
 *
 * Copyright 2015-2018 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.nadmm.airports;

import android.os.Bundle;

import com.google.android.material.tabs.TabLayout;
import com.nadmm.airports.utils.PagerAdapter;

import androidx.viewpager.widget.ViewPager;

public abstract class TabPagerActivityBase extends ActivityBase {

    private int mCurrentTabIndex = -1;
    private ViewPager mViewPager;
    private PagerAdapter mPagerAdapter;
    private TabLayout mTabLayout;

    private static final String SAVED_TAB = "saved_tab";

    @Override
    protected void onCreate( Bundle savedInstanceState ) {
        super.onCreate( savedInstanceState );

        setContentView( R.layout.activity_tab_pager );

        mViewPager = findViewById( R.id.view_pager );
        mPagerAdapter = new PagerAdapter( this, getSupportFragmentManager(), mViewPager );

        mViewPager.addOnPageChangeListener( new ViewPager.SimpleOnPageChangeListener() {

            public void onPageScrollStateChanged( int state ) {
                // Disable the swipe refresh while moving between pages
                if ( getCurrentFragment().isRefreshable() ) {
                    enableDisableSwipeRefresh( state == ViewPager.SCROLL_STATE_IDLE );
                }
            }
        } );

        mTabLayout = findViewById( R.id.sliding_tabs );
    }

    @Override
    protected void onPostCreate( Bundle savedInstanceState ) {
        super.onPostCreate( savedInstanceState );

        mTabLayout.setupWithViewPager( mViewPager );
        mTabLayout.addOnTabSelectedListener( new TabLayout.OnTabSelectedListener() {

            @Override
            public void onTabSelected( TabLayout.Tab tab ) {
                mCurrentTabIndex = tab.getPosition();
                mViewPager.setCurrentItem( mCurrentTabIndex );
                enableDisableSwipeRefresh( getCurrentFragment().isRefreshable() );
                showAppBar( true );
            }

            @Override
            public void onTabUnselected( TabLayout.Tab tab ) {
            }

            @Override
            public void onTabReselected( TabLayout.Tab tab ) {
                if ( mCurrentTabIndex != tab.getPosition() ) {
                    onTabSelected( tab );
                }
            }
        } );

        if ( savedInstanceState != null ) {
            mCurrentTabIndex = savedInstanceState.getInt( SAVED_TAB );
        } else {
            mCurrentTabIndex = getInitialTabIndex();
        }

        postRunnable( () -> {
            if ( mCurrentTabIndex >= 0 && mCurrentTabIndex < mTabLayout.getTabCount() ) {
                mTabLayout.getTabAt( mCurrentTabIndex ).select();
                enableDisableSwipeRefresh( getCurrentFragment().isRefreshable() );
            }
        }, 0 );
    }

    @Override
    public void onSaveInstanceState( Bundle outState ) {
        super.onSaveInstanceState( outState );

        outState.putInt( SAVED_TAB, mViewPager.getCurrentItem() );
    }

    @Override
    protected void requestDataRefresh() {
        FragmentBase fragment = getCurrentFragment();
        if ( fragment.isRefreshable() ) {
            fragment.requestDataRefresh();
        }
    }

    public void addTab( String label, Class<?> clss, Bundle args ) {
        mPagerAdapter.addTab( label, clss, args );
    }

    private FragmentBase getCurrentFragment() {
        return (FragmentBase) mPagerAdapter.getItem( mCurrentTabIndex );
    }

    protected int getInitialTabIndex() {
        return 0;
    }

}

File: src/main/java/com/nadmm/airports/TabPagerActivityBase.kt
/*
 * FlightIntel for Pilots
 *
 * Copyright 2015-2025 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.nadmm.airports

import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import com.nadmm.airports.utils.PagerAdapter

abstract class TabPagerActivityBase : ActivityBase() {
    private var activeTabIndex = -1
    private lateinit var fragmentPager: ViewPager2
    private lateinit var pagerAdapter: PagerAdapter
    private lateinit var tabs: TabLayout

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_tab_pager)
        fragmentPager = findViewById(R.id.view_pager)
        pagerAdapter = PagerAdapter(this)
        fragmentPager.adapter = pagerAdapter
        tabs = findViewById(R.id.sliding_tabs)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)

        TabLayoutMediator(tabs, fragmentPager) { tab, position ->
            tab.text = pagerAdapter.getPageTitle(position)
        }.attach()

        fragmentPager.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
            override fun onPageScrollStateChanged(state: Int) {
                // Disable the swipe refresh while moving between pages
                currentFragment?.let { fragment ->
                    if (fragment.isRefreshable) {
                        enableDisableSwipeRefresh(state == ViewPager2.SCROLL_STATE_IDLE)
                    }
                }
            }
        })

        fragmentPager.post {
            activeTabIndex = savedInstanceState?.getInt(SAVED_TAB) ?: initialTabIndex
            fragmentPager.setCurrentItem(activeTabIndex, false)
        }
    }

    public override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt(SAVED_TAB, fragmentPager.currentItem)
    }

    override fun requestDataRefresh() {
        currentFragment?.let { fragment ->
            if (fragment.isRefreshable) {
                fragment.requestDataRefresh()
            }
        }
    }

    protected fun addTab(label: String, clss: Class<out Fragment>, args: Bundle?=null) {
        pagerAdapter.addTab(label, clss, args)
    }

    private val currentFragment: FragmentBase?
        get() = pagerAdapter.getItem(fragmentPager.currentItem)

    protected open val initialTabIndex: Int = 0

    companion object {
        private const val SAVED_TAB = "saved_tab"
    }
}